src/main/java/com/example/autokis/parser/impl/ParserImpl.java

```diff
@@ -0,0 +1,167 @@
+package com.example.autokis.parser.impl;
+
+import com.example.autokis.parser.Parser;
+import com.example.autokis.parser.model.Product;
+import lombok.extern.slf4j.Slf4j;
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+import org.jsoup.select.Elements;
+import org.springframework.stereotype.Component;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+@Component("auto-elegant")
+@Slf4j
+public class ParserImpl implements Parser {
+    private static final String WEBSITE_MAIN_URL = "https://auto-elegant.com";
+    private Document currentDocument;
+    private static final String PAGE_PREFIX = "?page=";
+
+    private void connectToWebsite() throws IOException {
+        currentDocument = Jsoup.connect(WEBSITE_MAIN_URL).get();
+    }
+
+    public List<String> parseCategories() throws IOException {
+        connectToWebsite();
+        Elements linksToCategories = currentDocument.select("aside ul li a");
+        List<String> urls = new ArrayList<>();
+
+        for (Element e : linksToCategories) {
+            String validUrl = modifyUrlIfNotFull(e.attr("href"));
+            urls.add(validUrl);
+        }
+        return urls;
+    }
+
+    public List<Product> parseGoodsFromOneUrl(String url) throws IOException {
+        Document pageWithAllGoods = Jsoup.connect(url).get();
+        Element pagesString = pageWithAllGoods.getElementsByClass("results").first();
+        int pagesCountFromPageString = getPagesCountFromPageString(pagesString.text());
+        List<String> productLinks = new ArrayList<>();
+        List<Product> parsedProducts = new ArrayList<>();
+
+        for (int i = 1; i < pagesCountFromPageString + 1; i++) {
+            productLinks.addAll(parseSinglePage(url, i));
+        }
+
+        for (String link : productLinks) {
+            try {
+                parsedProducts.add(parseSingleProduct(link));
+            } catch (IOException e) {
+                log.warn("Error parsing product with link %s \n Exception: ".formatted(link), e);
+            }
+        }
+
+        return parsedProducts;
+    }
+
+    public Product parseSingleProduct(String url) throws IOException {
+        Document productDocument = Jsoup.connect(url).followRedirects(true).get();
```

Comment:
I would extract all Network operations out if this class to a separate class that can be injected here